### PR TITLE
Improve automated builds

### PR DESCRIPTION
**Pull request by cwillisf**
### Proposed Changes

- Support non-signed builds for PRs from forks
- Support automated macOS builds
- Determine whether or not to sign the build based on the branch
- Add a script to generate self-signed SSL certificates and use it for non-signed CI builds
   - This script is also useful for local development
- Share CircleCI config between Windows and macOS where possible

### Reason for Changes

This should bring Scratch Link's automated builds to a fully functional state. Specifically:

- We will automatically build any branch in LLK/scratch-link as well as pull requests targeted at any branch in LLK/scratch-link
- The branches `develop`, `main`, and `master` are treated specially:
   - These builds will use proper, official SSL certificates
   - The application and installer will be digitally signed
- In other branches, including PRs:
   - These builds will use self-signed "mock" SSL certificates
   - The application and installer will not be digitally signed

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,31 +1,181 @@
...
-#      - build_for_mac
+  macos:
+    jobs:
+      - build_for_mac:
+          context: scratch-desktop-and-link
   windows:
     jobs:
       - build_for_windows:
-          context: code-sign-generic
+          context: scratch-desktop-and-link
```
Comment (benjiwheeler) [inline .circleci/config.yml:10]: context is configured through Circle site; it's a batch of env vars. Assumption is that they are secret.

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,31 +1,181 @@
...
 orbs:
   windows: circleci/windows@2.2.0
+aliases:
+  - &should_sign
```
Comment (benjiwheeler) [inline .circleci/config.yml:14]: you should sign the build if you are on develop, main or master.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,31 +1,181 @@
...
+    or:
+      - equal: [ develop, << pipeline.git.branch >> ]
+      - equal: [ main, << pipeline.git.branch >> ]
+      - equal: [ master, << pipeline.git.branch >> ]
+commands:
+  prepare_certificates:
+    description: "This command prepares SSL certificates in the Certificates directory."
+    steps:
+      - unless:
```
Comment (benjiwheeler) [inline .circleci/config.yml:23]: these unless... when items act effectively as an if.. else

File: .circleci/config.yml [hunk 4]
```diff
@@ -1,31 +1,181 @@
...
+                name: Generate mock SSL certificates
+                working_directory: ./Certificates
+                shell: bash
+                command: ./mock-certificates.sh
+      - when:
+          condition:
+            *should_sign
+          steps:
+            - run:
+                name: Import from CI context
```
Comment (benjiwheeler) [inline .circleci/config.yml:39]: We need certificate files, but don't want those files checked into the repo. So we store their content in Base64 env vars, and decode those to local files. Those files will need further conversion, provided in convert-certificates.sh below.

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,31 +1,181 @@
...
+                # DO NOT RUN THIS on your own computer unless you really know what you're doing!
+                command: |
+                  security -v create-keychain -p circleci circleci.keychain
+                  security -v default-keychain -s circleci.keychain
+                  security -v import Certificates/code-to-learn-macos.p12 -k circleci.keychain -P "" -T /usr/bin/codesign -T /usr/bin/productbuild
+                  security -v unlock-keychain -p circleci circleci.keychain
+                  # "set-key-partition-list" prints extensive not-so-useful output and adding "-q" (even multiple times) doesn't suppress it.
+                  # The "grep -v" at the end of this line suppresses all of that so any errors or warnings might be more visible.
+                  security -v set-key-partition-list -S apple-tool:,apple:,codesign: -s -k circleci circleci.keychain | grep -v '^    0x'
+                  security -v set-keychain-settings -lut 600 circleci.keychain
```
Comment (benjiwheeler) [inline .circleci/config.yml:112]: stay unlocked for 10 minutes, enough time to build on circle

File: .circleci/config.yml [hunk 6]
```diff
@@ -1,31 +1,181 @@
...
+          key: homebrew-cache-v2
+      - run:
+          name: Install Homebrew packages
+          command: brew install pngcrush
+      - save_cache:
+          name: Save Homebrew cache
+          paths:
+            - ~/Library/Caches/Homebrew
+            - /usr/local/Homebrew
+          key: homebrew-cache-v2
```
Comment (benjiwheeler) [inline .circleci/config.yml]: important: you should manually bump this version number if you make changes, such as adding a path to cache

File: .circleci/config.yml [hunk 7]
```diff
@@ -1,31 +1,181 @@
...
+          key: spm-cache-v2-{{ checksum "macOS/Package.resolved" }}
+      - run:
+          name: Resolve Swift dependencies
+          command: make --directory=macOS resolve-dependencies
+      - save_cache:
+          name: Save Swift dependency cache
+          key: spm-cache-v2-{{ checksum "macOS/Package.resolved" }}
+          paths:
+            - macOS/.build
+      - run:
```
Comment (benjiwheeler) [inline .circleci/config.yml:138]: this just provides info to log on which files git considers changed and not staged -- helpful for investigating/debugging

File: .circleci/config.yml [hunk 8]
```diff
@@ -1,31 +1,181 @@
...
+            - run:
+                name: Import code-signing certificate
+                shell: powershell
+                command: |
+                  $ErrorActionPreference = "Stop"
+                  # TODO: is there a way to avoid creating a temporary file?
+                  $tempPfx = New-TemporaryFile
+                  try {
+                      $env:WIN_CSC_LINK | Set-Content $tempPfx
+                      $securePassword = (ConvertTo-SecureString -String $env:WIN_CSC_KEY_PASSWORD -AsPlainText -Force)
```
Comment (benjiwheeler) [inline .circleci/config.yml:188]: password below needs to be a SecureString, so convert our password to one

File: .circleci/config.yml [hunk 9]
```diff
@@ -1,31 +1,181 @@
...
+                      Import-PfxCertificate -FilePath $tempPfx -Password $securePassword -CertStoreLocation "Cert:/LocalMachine/My"
```
Comment (benjiwheeler) [inline .circleci/config.yml:189]: Import certificate and store in Windows' Certificate store

File: .circleci/config.yml [hunk 10]
```diff
@@ -1,31 +1,181 @@
...
+          paths:
+            - ~/Library/Caches/Homebrew
+            - /usr/local/Homebrew
+          key: homebrew-cache-v2
+      - restore_cache:
+          name: Restore Swift dependency cache
+          key: spm-cache-v2-{{ checksum "macOS/Package.resolved" }}
+      - run:
+          name: Resolve Swift dependencies
+          command: make --directory=macOS resolve-dependencies
```
Comment (cwillisf) [inline .circleci/config.yml:132]: In theory these three sequences could be run in parallel:
- prepare_certificates / set up code-sign keychain
- restore Swift dependency cache / resolve Swift dependencies / save Swift dependency cache
- restore Homebrew cache / install pngcrush / save Homebrew cache

They should all finish before the "Check git status" step is started, though, and steps after that point must run in serial.

**Pull request merged**